### PR TITLE
Make sure circuits with constant ops can be serialized to JSON correctly

### LINKED ISSUE #65: Cannot save simple design due to BigInt
The following simple design cannot be saved due to JSON.stringify() not knowing how to serialize a `BigInt`.

```
module counter(input wire clock, output reg [0:0] out);
   initial begin
      out = 1'b0;
   end

   always @(posedge clock) begin
      out <= out + 1'b1;
   end
endmodule
```

The culprit seems to be the increment. Monkey patching BigInt using https://github.com/GoogleChromeLabs/jsbi/issues/30#issuecomment-953187833 seems to work around the issue and produces a usable json output. A less invasive method by adding a replacement argument to the stringify call is probably better.

I would have tested this locally and submit a PR instead but somehow sass dependency doesn't work on node 17 right now so I couldn't test this...

### PR DESCRIPTION
**Pull request by yuyichao**
The fix is limited to OpConst to make sure we can handle the deserialization without any problems. Also fixed the implementation of ShiftConst to handle BigInt produced by constant folding.

Add tests to check the validity and stability of the JSON object.

Fix #65

.... took me longer to add the tests than to fix the issue....

### CODE REVIEW

File: src/cells/arith.mjs [hunk 1]
```diff
@@ -368,6 +368,13 @@ export const OpConst = Arith.define('OpConst', {
     initialize() {
         Arith.prototype.initialize.apply(this, arguments);
 
+        let constant = this.get('constant');
+        if (typeof constant === 'string' || constant instanceof String) {
+            constant = BigInt(constant);
+            if (constant <= BigInt(Number.MAX_SAFE_INTEGER))
```
Comment (tilk) [inline src/cells/arith.mjs:374]: What about negative numbers?

File: src/cells/arith.mjs [hunk 2]
```diff
@@ -384,6 +391,19 @@ export const OpConst = Arith.define('OpConst', {
         this.on('change:constant', () => genLabel());
         this.on('change:leftOp', () => genLabel());
     },
+    getGateParams() {
+        const params = Arith.prototype.getGateParams.apply(this, arguments);
+        // `constant` may be a bigint and we need to encode it manually
+        if (typeof params.constant === 'bigint') {
+            if (params.constant <= BigInt(Number.MAX_SAFE_INTEGER)) {
```
Comment (tilk) [inline src/cells/arith.mjs:398]: Negative numbers.

File: tests/index.test.mjs [hunk 3]
```diff
@@ -366,6 +476,34 @@ describe.each([
...
+                    .testJSON();
+                const test2 = new SingleCellConstTestFixture(engine, {type: name, bits: { in1: bits, in2: bits, out: bits }, signed: { in1: sgn, in2: sgn }}, {input: 'in2', bits: conbin})
+                    .testJSON();
+                if (testval) {
+                    test1.testFun(s => fun(sgn, sgn, bits)({in1: conbits, in2: s.in2}), { no_random_x : true });
+                    test2.testFun(s => fun(sgn, sgn, bits)({in1: s.in1, in2: conbits}), { no_random_x : true });
+                }
+            };
+            // Make sure we test both Number and BigInt
+            testcon(randInt(0, Math.min(2**bits - 1, Number.MAX_SAFE_INTEGER)));
```
Comment (tilk) [inline tests/index.test.mjs:501]: Negative numbers are not tested.

**Review by tilk — CHANGES_REQUESTED** [review]
One more comment: if you look at `integrateArithConstant`, you'll see it only generates *Const cells for "small" constants - which was done for diagram readability reasons. But the constant is always of BigInt type, and it could already be a Number there.

**Comment by yuyichao** [comment]
Rebased and fixed.

This now eagerly convert the constant to `Number` when doing the transformation but it will still accept bigint if explicitly constructed.

**Pull request merged**